Prevent logstash from running as root

mashhurs opened the pull request:
<!-- Type of change
Please label this PR with the release version and one of the following labels, depending on the scope of your change:
- bug
- enhancement
- breaking change
- doc
-->

## Release notes
<!-- Add content to appear in  [Release Notes](https://www.elastic.co/guide/en/logstash/current/releasenotes.html), or add [rn:skip] to leave this PR out of release notes -->


## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.

Example:
  Expose 'xpack.monitoring.elasticsearch.proxy' in the docker environment variables and update logstash.yml to surface this config option.
  
  This commit exposes the 'xpack.monitoring.elasticsearch.proxy' variable in the docker by adding it in env2yaml.go, which translates from
  being an environment variable to a proper yaml config.
  
  Additionally, this PR exposes this setting for both xpack monitoring & management to the logstash.yml file.
-->

## Why is it important/What is the impact to the user?

<!-- Mandatory
Explain here the WHY or the IMPACT to the user, or the rationale/motivation for the changes.

Example:
  This PR fixes an issue that was preventing the docker image from using the proxy setting when sending xpack monitoring information.
  and/or
  This PR now allows the user to define the xpack monitoring proxy setting in the docker container.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

Cases:
1. Unintentionally running Logstash as a root
 - run_as.root: false in logstash.yaml
 - run with sudo ./bin/logstash
 - Output: [2022-04-22T22:04:36,115][FATAL][logstash.runner] An unexpected error occurred! {:error=>#<RuntimeError: can not run logstash as root>, :backtrace=>["/Users/mashhur/Dev/elastic/logstash/logstash-core/lib/logstash/runner.rb:292:in `execute'",

2. Intentionally running Logstash as a root
 - run_as.root: true in logstash.yaml
 - run with sudo ./bin/logstash
 - Output: [2022-04-22T22:16:32,100][WARN ][logstash.runner] running logstash as root is not recommended because of security reasons

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseeds #123
-->
- Closes #13882

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->
```
// unintentionally run as root
[2022-04-22T22:04:36,115][FATAL][logstash.runner] An unexpected error occurred! {:error=>#<RuntimeError: can not run logstash as root>, :backtrace=>["/Users/mashhur/Dev/elastic/logstash/logstash-core/lib/logstash/runner.rb:292:in `execute'",
```
```
// intentionally run as root
[2022-04-22T22:16:32,100][WARN ][logstash.runner          ] running logstash as root is not recommended because of security reasons
```

mashhurs commented:
I have following unit test cases but no idea how to mock `Process.euid()`. Can anyone has an idea?

```
  describe "run_as.root" do
    subject { LogStash::Runner.new("") }

    context "unintentionally running logstash as root" do
      let(:settings) {
        {"run_as.root" => false }
      }
      it "fails with bad exit" do
        expect(logger).to receive(:fatal)
        expect(subject.run(args)).to eq(1)
      end
    end

    context "intentionally running logstash as root " do
      let(:settings) {
        {"run_as.root" => true }
      }
      it "successfully runs with warning message" do
        expect(logger).not_to receive(:fatal)
        expect(subject.run(args)).to eq(0)
      end
    end
  end
```

mashhurs commented:
@jsvd, @robbavey can you please let me know how to test on docker & Windows?
I also do see we have separate separate `env2yaml.go` file for docker but I have limited visibility how currently we run Logstash on docker. Thanks.

karenzone commented on config/logstash.yml:
```diff
@@ -281,6 +281,9 @@
 # ------------ Other Settings --------------
 #
+# Run Logstash with root
+run_as.root: true
```
This seems backwards to me.  If we're preventing Logstash from running as root, it seems like the default should be `false`.

yaauie replied:
The `config/logstash.yml` we ship should be 100% comments, matching the defaults defined in `environment.rb`.

mashhurs replied:
@karenzone let's discuss doc sections in our meeting (on April 27th) and as I mentioned in weekly sync up, initially we offer `NOTICE (non-breaking)` option and let users know that in the future option will be changed (to `BLOCK`).

karenzone commented:
General best practice for computing is to NOT run as root, so I don't think we need to make a big deal about this.  You have messaging in place, just in case somebody tries. 

We could consider adding a short topic to Troubleshooting doc, just to reinforce that running as root in not a good idea, and in fact, is not allowed. Troubleshooting docs start by identifying error condition, and then providing additional detail. Something like that could go in the Installation and setup section, either before or after the "inaccessible temp directory. topic: https://github.com/elastic/logstash/blame/main/docs/static/troubleshoot/ts-logstash.asciidoc#L6

karenzone commented on logstash-core/lib/logstash/runner.rb:
```diff
@@ -285,6 +285,12 @@ def execute
 
+    if Process.euid() == 0
+      run_as_root = setting("run_as.root") == true
+      raise(RuntimeError,"can not run logstash as root") if !run_as_root
```
```suggestion
      raise(RuntimeError,"Cannot run logstash as root") if !run_as_root
```

karenzone replied:
Or perhaps "Logstash cannot be run as root"

mashhurs replied:
This suggestion LGTM!

karenzone commented on logstash-core/lib/logstash/runner.rb:
```diff
@@ -285,6 +285,12 @@ def execute
+    if Process.euid() == 0
+      run_as_root = setting("run_as.root") == true
+      raise(RuntimeError,"can not run logstash as root") if !run_as_root
+      logger.warn("running logstash as root is not recommended because of security reasons") if run_as_root
```
If we're preventing Logstash from running as root, the error message should reflect that rather than saying it's not recommended. 

Something like, "Logstash cannot be run as root for security reasons

mashhurs replied:
This suggestion LGTM!

karenzone reviewed (commented):
Left some comments and suggestions, and some ideas on how to approach documentation.  Happy to discuss if you'd like.

roaksoax commented on logstash-core/lib/logstash/environment.rb:
```diff
@@ -34,6 +34,7 @@ module Environment
   end
 
   [
+           Setting::Boolean.new("run_as.root", false),
```
I find it a bit weird we define the setting as 'run_as.root'. I doubt we will ever have to add other settings under 'run_as', so I would think we should just create a flat config name, e.g.:

'allow_root'

Or similar.

mashhurs replied:
Indeed naming is always difficult :) We shortly discussed with @yaauie and making it `open to extend` instead of limiting to `boolean` makes sense for this case. And finalized with following config:
`on_superuser: BLOCK, WARN(default), ALLOW`

Let me please know if still I can improve or missing. Thanks!

yaauie commented:
Implementation makes sense.

I would like to step back and look at the interface.
 - `run_as.root` in a `logstash.yml` config file to me implies that we will attempt to run the process as root, which isn't what we do
 - `allow_root`, as suggested in this PR forces a boolean option. I prefer avoiding booleans because inevitably there ends up being a third option introduced (such as the ability to mute this warning, for users who knowingly run as root in a controlled environment where root is limited). We also run into problems with the terminology `root` when Logstash is run on Windows (where we may want to limit being run-as Administrator in a future release).

Perhaps we can add a setting `on_superuser` (or similar) that takes one of three settings:
 - `BLOCK`: raises a helpful error.
 - `WARN` (default): warns a helpful warning.
 - `ALLOW`: silently allows, or avoids checking.

This setting could be documented as known working on POSIX platforms, and could be extended to include Windows at a point in the future.

jsvd reviewed (commented):
A few notes:

**Future: never allow running as root**

While we're just introducing this feature, I'd suggest making it deprecated with a notice that we will not allow it to run as root in the future, so BLOCK will eventually be the default.
Users should be paying attention to their deprecation logs and setting the value to BLOCK so they don't have issues in the future.
For users running as root for..reasons, I'm sure there are much safer solutions. For example, if there's a need to bind ports <1024, users can leverage either port forwarding, authbind, or even the CAP_NET_BIND_SERVICE capability.

**Missing exposure of `on_superuser`**

The new setting should be added to `./docker/data/logstash/env2yaml/env2yaml.go` so users can set it in docker images through environment variables.

The docs file `./docs/static/settings-file.asciidoc` lists all possible settings, so it should also have a section on `on_superuser`.

mashhurs commented:
> A few notes:
> 
> **Future: never allow running as root**
> 
> While we're just introducing this feature, I'd suggest making it deprecated with a notice that we will not allow it to run as root in the future, so BLOCK will eventually be the default. Users should be paying attention to their deprecation logs and setting the value to BLOCK so they don't have issues in the future. For users running as root for..reasons, I'm sure there are much safer solutions. For example, if there's a need to bind ports <1024, users can leverage either port forwarding, authbind, or even the CAP_NET_BIND_SERVICE capability.
> 
> **Missing exposure of `on_superuser`**
> 
> The new setting should be added to `./docker/data/logstash/env2yaml/env2yaml.go` so users can set it in docker images through environment variables.
> 
> The docs file `./docs/static/settings-file.asciidoc` lists all possible settings, so it should also have a section on `on_superuser`.

Thanks @jsvd,
Added docker a variable in docker environment and run both docker & windows acceptance tests.
Also, added documentation and I scheduled a meeting with Karen to double check if all messages make sense.
Thanks.

jsvd commented on logstash-core/lib/logstash/environment.rb:
```diff
@@ -34,6 +34,7 @@ module Environment
   end
 
   [
+            Setting::String.new("on_superuser", "WARN"),
```
Similarly to other settings:

```
            Setting::String.new("log.level", "info", true, ["fatal", "error", "warn", "debug", "info", "trace"]),
```

 we can list the acceptable values in the setting definition:

```suggestion
            Setting::String.new("on_superuser", "WARN", true, ["ALLOW", "WARN", "BLOCK"]),
```

jsvd commented on logstash-core/lib/logstash/runner.rb:
```diff
@@ -440,6 +442,21 @@ def execute
 
+  def running_as_root
+    running_as_root = setting("on_superuser")
+    running_as_root_allowed_options = { "allow" => "ALLOW", "warn" => "WARN", "block" => "BLOCK" }
+    if running_as_root_allowed_options.has_value?(running_as_root) == false
+      logger.warn("Allowed options on on_superuser are WARN, BLOCK and ALLOW. Falling back to default WARN.")
+      running_as_root = running_as_root_allowed_options["warn"]
+    end
```
If we set the list of possible values in the setting definition we can do away with this code.

mashhurs replied:
Thanks!
I also figured out that if I add `on_superuser ` with options in `environment.rb` declaration, we don't need this lines.
```
Setting::String.new("on_superuser", "WARN", true, ["WARN", "BLOCK", "ALLOW"]),
```

jsvd commented on logstash-core/lib/logstash/runner.rb:
```diff
@@ -440,6 +442,21 @@ def execute
+      running_as_root = running_as_root_allowed_options["warn"]
+    end
+    if Process.euid() == 0 && running_as_root.eql?(running_as_root_allowed_options["allow"]) == false
+      deprecation_message_on_root = "NOTICE: Running logstash as root is not recommended because of security reasons. This will be deprecated in the future."
+      error_message_on_root = "Logstash cannot be run as root."
+      logger.warn(deprecation_message_on_root) if running_as_root.eql?(running_as_root_allowed_options["warn"])
+      raise(RuntimeError, error_message_on_root) if running_as_root.eql?(running_as_root_allowed_options["block"])
```
We can log a deprecation log using the deprecation logger instead `logger.warn`. This way the logging entry ends up the right log file (i.e. `logs/logstash-deprecation.log` vs `logs/logstash-plain.log`).
Also we should add a call to action in the deprecation notice to guide to user towards what they need to do:

```suggestion
      if running_as_root == "WARN"
        deprecation_logger.deprecated("NOTICE: Running Logstash as root is not recommended and won't be allowed in the future. Set 'on_superuser' to 'BLOCK' to avoid startup errors in future releases.".)
      else # BLOCK
        raise(RuntimeError, "Logstash cannot be run as root.") 
      end
```

mashhurs replied:
I have limited visibility to `deprecation logs` and one question though.
I don't see any types of logs are appeared on console when I use `deprecation_logger` and message will be logged into `logstash-deprecation.log` file. I wonder how customers are able to easily see this log to take an action?

jsvd commented on logstash-core/lib/logstash/environment.rb:
```diff
@@ -34,6 +34,7 @@ module Environment
   end
 
   [
+            Setting::String.new("on_superuser", "WARN"),
```
Since we'll want to guide users towards not using root, instead of these three options we could instead do:
1. (default) allow but log deprecation warn
2. block

With only two options we can instead have a boolean setting:

```suggestion
            Setting::Boolean.new("allow_superuser", true),
```

mashhurs replied:
Thanks @jsvd for the suggestion.
Initially had a same thought. We with @yaauie had a discussion on this and we concluded that option should be `open to extend` in the future. You can take a look at his comments under this PR and let me know if you have still concern.
Thanks.

jsvd reviewed (commented):
A few suggestions:

1. use deprecation_logger instead of the normal logger
2. reduce options to two, making the setting boolean: either allow and log, or block.
